keps/sig-storage/1702-volume-provisioning-rollback/README.md

```diff
@@ -0,0 +1,346 @@
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+When a pod uses one or more volumes with late binding (aka
+`WaitForFirstConsumer`), volume provisioning is kicked of when a pod
+gets tentatively assigned to a node. When one of the volumes cannot be
+provisioned for the selected node, the scheduler needs to try with
+some other node. But it cannot do that if the already provisioned
+volumes are only available on the original node.
+
+The KEP proposes a rollback mechanism that undos the provisioning of
+volumes that block the rescheduling of a pod.
+
+## Motivation
+
+Storage capacity aware pod scheduling increases the chance that a node
+is picked which still has enough capacity left. However, there are
+still cases (race conditions, insufficient information) where the
+actual attempts to create volumes fail.
+
+If scheduling was based on stale information, the next attempt may be
+able to pick a better node. However for that to work, Kubernetes has
+to prevent getting stuck with partially allocated resources for the
+pod.
+
+### Goals
+
+- Enable recovery from a bad node selection for a pod when volume
+  provisioning fails for some, but not all nodes.
+
+### Non-Goals
+
+- Only CSI drivers using an up-to-date external-provisioner will be
+  supported.
+
+## Proposal
```

Comment:
I would prefer having some stronger guarantee on provisioning and rolling back a group of volumes atomically to avoid races that could result in deletion of a volume in use.

So something that can prevent a Pod from using a volume before all the volumes are successfully provisioned. Maybe something like taints can help us here.

Reply:
>> I would prefer having some stronger guarantee on provisioning and rolling back a group of volumes atomically
> As multiple different storage systems are involved, it will have to be Kubernetes which treats the set of volumes as something that is kept unused until all volumes are ready.



> > I would prefer having some stronger guarantee on provisioning and rolling back a group of volumes atomically
> 
> As multiple different storage systems are involved, it will have to be Kubernetes which treats the set of volumes as something that is kept unused until all volumes are ready.

As for static binding, we need to reverse multiple volumes for later use too (https://github.com/kubernetes/kubernetes/issues/73263). It's a special case in which volumes are already provisioned and don't need provisioning rollback.